aws-datastore/src/main/java/com/amplifyframework/datastore/network/SubscriptionProcessor.java

```diff
@@ -0,0 +1,126 @@
+            @NonNull ModelProvider modelProvider) {
+        this.localStorageAdapter = localStorageAdapter;
+        this.remoteModelMutations = new RemoteModelMutations(appSyncEndpoint, modelProvider);
+        this.disposable = new CompositeDisposable();
+        this.buffer = ReplaySubject.create();
+    }
+
+    /**
+     * Start subscribing to model mutations.
+     */
+    void startSubscriptions() {
+        disposable.add(remoteModelMutations.observe()
+            .subscribe(
+                mutationOnSubscription -> {
+                    buffer.onNext(mutationOnSubscription);
+                    LOG.info("Successfully enqueued mutation from subscription: " + mutationOnSubscription);
+                },
+                error -> LOG.warn("Error enqueuing mutation from subscription.", error),
+                () -> LOG.warn("Subscription to remote model mutations is completed.")
+            )
+        );
+    }
+
+    /**
+     * Start draining mutations out of the mutation buffer.
+     * This should be called after {@link #startSubscriptions()}.
+     */
+    void startDrainingMutationBuffer() {
+        disposable.add(buffer
+            .flatMapSingle(this::applyMutationToLocalStorage)
+            .subscribe(
+                savedMutation -> LOG.info("Saved a mutation from a subscription: " + savedMutation),
+                failure -> LOG.warn("Reading subscriptions buffer has failed.", failure),
+                () -> LOG.warn("Reading from subscriptions buffer is completed.")
+            )
+        );
+    }
+
+    /**
+     * Stop any active subscriptions, and stop draining the mutation buffer.
+     */
+    void stopAllSubscriptionActivity() {
+        disposable.clear();
+    }
+
+    private Single<Mutation<? extends Model>> applyMutationToLocalStorage(Mutation<? extends Model> mutation) {
+        final StorageItemChange.Initiator initiator = StorageItemChange.Initiator.SYNC_ENGINE;
+        return Single.defer(() -> Single.create(emitter -> {
+            final Consumer<StorageItemChange.Record> onSuccess =
+                result -> emitter.onSuccess(mutation);
+            final Consumer<DataStoreException> onError = emitter::onError;
+
+            switch (mutation.type()) {
+                case UPDATE:
+                case CREATE:
+                    localStorageAdapter.save(mutation.model(), initiator, onSuccess, onError);
+                    break;
+                case DELETE:
+                    localStorageAdapter.delete(mutation.model(), initiator, onSuccess, onError);
+                    break;
+                default:
+                    throw new DataStoreException(
+                        "Unknown mutation type = " + mutation.type(),
+                        AmplifyException.TODO_RECOVERY_SUGGESTION
+                    );
+            }
+
+            // Notify Hub that we just updated the local storage.
+            HubEvent<? extends Model> receivedFromCloudEvent =
```

Comment:
Will this get really noisy? I had suggested notifications from subscriptions to Hub in the past and the pushback was that the messages might get really verbose if you get a lot of subscription updates.

Reply:
Eh, yea. Good catch, and good call out. One of these events for every subscription event is probably going to get unwieldly ...

Maybe we can add an off/on switch in the config - default is on, and then if the customer is like "whoa, are you kidding me?" our documentation does mention a way to shut it down?